File: cordova-plugin-appcenter-analytics/src/android/AppCenterAnalyticsPlugin.java
package com.microsoft.azure.mobile.cordova;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import com.microsoft.appcenter.AppCenter;
import com.microsoft.appcenter.analytics.Analytics;

public class AppCenterAnalyticsPlugin extends CordovaPlugin {
    private static final String ENABLE_IN_JS = "APPCENTER_ANALYTICS_ENABLE_IN_JS";

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);

        AppCenterShared.configureAppCenter(
                cordova.getActivity().getApplication(),
                webView.getPreferences());

        // TODO: once the underlying SDK supports this, make sure to call this
        //Analytics.setAutoPageTrackingEnabled(false);
        AppCenter.start(Analytics.class);

        boolean enableInJs = webView.getPreferences().getBoolean(ENABLE_IN_JS, false);
        if (enableInJs) {
            // Avoid starting an analytics session.
            // Note that we don't call this if startEnabled is true, because
            // that causes a session to try and start before Analytics is started.
            Analytics.setEnabled(false);
        }
    }

    @Override
    public boolean execute(String action, JSONArray args,
                           final CallbackContext callbackContext) throws JSONException {

        if (action.equals("trackEvent")) {

            String eventName = args.getString(0);
            JSONObject properties = args.getJSONObject(1);

            try {
                Analytics.trackEvent(eventName, AnalyticsUtils.stringMapFromJsonObject(properties));
                callbackContext.success();
            } catch (JSONException e) {
                callbackContext.sendPluginResult(
                        new PluginResult(PluginResult.Status.JSON_EXCEPTION, e.getMessage()));
            }

            return true;

        } else if (action.equals("isEnabled")) {
            AppCenterUtils.sendBooleanPluginResultFromFuture(Analytics.isEnabled(), callbackContext);
            return true;
        } else if (action.equals("setEnabled")) {
            Boolean enabled = args.getBoolean(0);
            AppCenterUtils.sendVoidPluginResultFromFuture(Analytics.setEnabled(enabled), callbackContext);
            return true;
        }

        return false;
    }
}

File: cordova-plugin-appcenter-analytics/src/ios/AppCenterAnalyticsPlugin.m
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.


#import <Cordova/NSDictionary+CordovaPreferences.h>

#import "AppCenterAnalyticsPlugin.h"
#import "AppCenterShared.h"

@import AppCenterAnalytics;

@implementation AppCenterAnalyticsPlugin

- (void)pluginInitialize
{
    [AppCenterShared configureWithSettings: self.commandDelegate.settings];
    [MSAppCenter startService:[MSAnalytics class]];

    BOOL enableInJs = [self.commandDelegate.settings
                       cordovaBoolSettingForKey:@"APPCENTER_ANALYTICS_ENABLE_IN_JS"
                       defaultValue:NO];

    if (enableInJs) {
        // Avoid starting an analytics session.
        // Note that we don't call this if startEnabled is true, because
        // that causes a session to try and start before MSAnalytics is started.
        [MSAnalytics setEnabled:false];
    }

    //[MSAnalytics setAutoPageTrackingEnabled:false]; // TODO: once the underlying SDK supports this, make sure to call this
}

- (void)isEnabled:(CDVInvokedUrlCommand *)command
{
    BOOL isEnabled = [MSAnalytics isEnabled];
    CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK
                                                  messageAsBool:isEnabled];

    [self.commandDelegate sendPluginResult:result
                                callbackId:command.callbackId];
}

- (void)setEnabled:(CDVInvokedUrlCommand *)command
{
    BOOL shouldEnable = [[command argumentAtIndex:0] boolValue];
    [MSAnalytics setEnabled:shouldEnable];

    CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:result
                                callbackId:command.callbackId];
}

- (void)trackEvent:(CDVInvokedUrlCommand *)command
{
    NSString* eventName = [command argumentAtIndex:0 withDefault:nil andClass:[NSString class]];
    NSDictionary* properties = [command argumentAtIndex:1];

    [MSAnalytics trackEvent:eventName withProperties:properties];

    CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:result
                                callbackId:command.callbackId];
}

/*
// TODO: once the underlying SDK supports this
- (void)trackEvent:(CDVInvokedUrlCommand *)command
{
    NSString pageName = [command argumentAtIndex:0];
    NSDictionary properties = [command argumentAtIndex:1];

    [MSAnalytics trackPage:pageName withProperties:properties];

    [self.commandDelegate sendPluginResult:CDVCommandStatus_OK
                                callbackId:command.callbackId];
}
*/

@end
